src/responder/ifp/ifp_users.c

```diff
@@ -1029,6 +1031,339 @@ ifp_users_list_by_domain_and_name_recv(TALLOC_CTX *mem_ctx,
+            DEBUG(SSSDBG_OP_FAILURE,
+                  P11_CHILD_PATH " failed with status [%d]\n", child_status);
+        } else {
+            valid = true;
+        }
+    }
+
+    if (WIFSIGNALED(child_status)) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              P11_CHILD_PATH " was terminated by signal [%d]\n",
+              WTERMSIG(child_status));
+    }
+
+    if (valid) {
+        DEBUG(SSSDBG_TRACE_LIBS, "Certificate [%s] is valid.\n",
+                                  state->extra_args[0]);
+
+        subreq = cache_req_user_by_cert_send(state, state->ifp_ctx->rctx->ev,
+                                             state->ifp_ctx->rctx,
+                                             state->ifp_ctx->rctx->ncache, 0,
+                                             CACHE_REQ_ANY_DOM, NULL, state->derb64);
+        if (subreq == NULL) {
+            DEBUG(SSSDBG_CRIT_FAILURE, "Unable to create subrequest!\n");
+            ret = ENOMEM;
+            goto done;
+        }
+
+        tevent_req_set_callback(subreq, ifp_users_find_by_valid_cert_step, req);
+    } else {
+        DEBUG(SSSDBG_MINOR_FAILURE, "Certificate [%s] is not valid.\n",
+                                    state->extra_args[0]);
+        tevent_req_error(req, ERR_INVALID_CERT);
+    }
+
+    ret = EAGAIN;
+
+done:
+    if (ret != EAGAIN) {
+        tevent_req_error(req, ret);
+        tevent_req_post(req, state->ifp_ctx->rctx->ev);
+    }
+
+    return req;
+}
+
+static void ifp_users_find_by_valid_cert_step(struct tevent_req *subreq)
+{
+    struct ifp_users_find_by_valid_cert_state *state;
+    struct cache_req_result *result;
+    struct tevent_req *req;
+    errno_t ret;
+
+    req = tevent_req_callback_data(subreq, struct tevent_req);
+    state = tevent_req_data(req, struct ifp_users_find_by_valid_cert_state);
+
+    ret = cache_req_user_by_cert_recv(state, subreq, &result);
+    talloc_zfree(subreq);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_MINOR_FAILURE, "Unable to find user [%d]: %s\n",
+              ret, sss_strerror(ret));
+        tevent_req_error(req, ret);
+        return;
+    }
+
+    if (result->count > 1) {
+         DEBUG(SSSDBG_CRIT_FAILURE, "More than one user found. "
+               "Use ListByCertificate to get all.\n");
+         tevent_req_error(req, EINVAL);
+         return;
+    }
+
+    state->path = ifp_users_build_path_from_msg(state, result->domain,
+                                                result->msgs[0]);
+    if (state->path == NULL) {
+        tevent_req_error(req, ENOMEM);
+        return;
+    }
+
+    tevent_req_done(req);
+    return;
+}
+
+errno_t
+ifp_users_find_by_valid_cert_recv(TALLOC_CTX *mem_ctx,
+                                  struct tevent_req *req,
+                                  const char **_path)
+{
+    struct ifp_users_find_by_valid_cert_state *state = tevent_req_data(req,
+                                                struct ifp_users_find_by_valid_cert_state);
```

Comment:
It's usually nicer to write:

```c
struct ifp_users_find_by_valid_cert_state *state;

state = tevent_req_data(req, struct ifp_users_find_by_valid_cert_state);
```
since it helps you fit 80 chars limit nicely.